### PR TITLE
TRUNK-4730 Verify that all relevant domain objects have changed_by and date_changed fields

### PR DESCRIPTION
**Pull request by manuela-grindei**
## Description

Added new attributes to classes, added new mappings to Hibernate files, added corresponding Liquibase patches
## Related Issue

see https://issues.openmrs.org/browse/TRUNK-4730

## Checklist:

- [x] My pull request only contains one single commit.
- [x] My pull request is based on the latest master branch
  `git pull --rebase upstream master`.
- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by dkayiwa** [comment]
@manuela-grindei thanks again for working on this! 👍 
Could you fix the integration test failure that happen with these changes?
To see the failures, just run: mvn clean install -Pintegration-test

File: api/src/main/resources/liquibase-update-to-2.0.xml [hunk 1]
```diff
@@ -9535,4 +9535,230 @@
         </sql>
     </changeSet>
 
+	<changeSet id="TRUNK-4730-20161114-1000" author="grindei" dbms="mysql">
+		<preConditions onFail="MARK_RAN">
+			<not>
+				<columnExists tableName="concept_class" columnName="date_changed"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase-update-to-2.0.xml]: Since the fix version is 2.1, can you move these changesets to liquibase-update-to-2.1.xml

Comment (manuela-grindei) [inline api/src/main/resources/liquibase-update-to-2.0.xml]: Done 👍

**Comment by manuela-grindei** [comment]
Fixed integration test failures.

**Comment by dkayiwa** [comment]
@manuela-grindei can you squash these commits into one as advised at https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by manuela-grindei** [comment]
Squashed commits

File: api/src/main/java/org/openmrs/BaseCustomizableData.java [hunk 2]
```diff
@@ -83,7 +83,6 @@ public void addAttribute(A attribute) {
 		if (getAttributes() == null) {
 			setAttributes(new LinkedHashSet<A>());
 		}
-		// TODO validate
```
Comment (manuela-grindei) [inline api/src/main/java/org/openmrs/BaseCustomizableData.java]: I had to squash everything, including a merge of upstream master, which gave those TODO removals.

File: api/src/test/resources/apiCacheConfig.properties [hunk 3]
```diff
@@ -0,0 +1,10 @@
+subscription.maxElementsInMemory=500
+subscription.eternal=false
+subscription.timeToIdleSeconds=300
+subscription.timeToLiveSeconds=300
```
Comment (dkayiwa) [inline api/src/test/resources/apiCacheConfig.properties]: Is this file also part of this ticket?

File: api/src/test/java/org/openmrs/api/cache/OpenmrsCacheManagerFactoryBeanTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/cache/OpenmrsCacheManagerFactoryBeanTest.java]: Is this part of this ticket?

File: api/src/main/java/org/openmrs/BaseCustomizableData.java (hunk 2 above)
Comment (manuela-grindei) [inline api/src/main/java/org/openmrs/BaseCustomizableData.java]: There used to be 3 commits, one of which was a master upstream merge

**Comment by dkayiwa** [comment]
@manuela-grindei i see so many changes that have nothing to do with this ticket. Can you fix that?

**Comment by manuela-grindei** [comment]
I'm afraid I don't know how...

**Comment by dkayiwa** [comment]
Did you get a chance to read this https://wiki.openmrs.org/display/docs/Pull+Request+Tips and https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by manuela-grindei** [comment]
Absolutely, I read it

**Comment by dkayiwa** [comment]
As a last resort, if you do not find it easy to separate your changes from other commits that have already been merged, you may just create a new branch and put back your changes one by one before making a new pull request.

**Comment by manuela-grindei** [comment]
This looks very strange, I think I will need to delete the repo again and recreate everything using patches... However, we need to merge quickly next time, so that we do not need to squash anything again. Please let me know if you are fine with the actual changes, as we need the final version in the next PR to be able to quickly merge it into master. We could choose a time of very low activity in the repo, maybe during Christmas, to prevent upstream master changes from happening.

**Comment by dkayiwa** [comment]
@manuela-grindei there will always be a possibility of commits before your pull request is merged. Even if a pull request is open and reviewed in 30 minutes, there will always be a chance of commits that happen so within for instance 15 minutes before the merging. This is the nature of distributed development. :)
You could choose a time of low activity when the activity is so low to even have someone to review and merge your changes. :)
So the ultimate solution is frequently pulling upstream changes into your local ones and getting to master how to squash commits.
In the ideal world, we would merge pull requests immediately, but this is not always possible because of lots of other pull requests we need to test and review, amidst other activities.
Even after a review which may look like the final version, trying to test out your changes by running the web application may prove otherwise. :)

**Comment by manuela-grindei** [comment]
You're totally right in what you say. 
I don't know why squashing commits doesn't work properly on my machine, I'll need to investigate and fix it, otherwise I'll be unable to contribute to openmrs.

**Comment by manuela-grindei** [comment]
In the meantime, I can provide a patch with the changes to you or someone else who can squash without issues, so we can still finish the story and merge the code in the repo.

**Comment by dkayiwa** [comment]
@manuela-grindei if you fail to get it right, you can post on talk, detailing the exact steps or commands you executed and the results. Then you will get answers from other contributors who have  been there before. For i have seem many of them successfully squash their commits. So be sure that you will get help from others' experiences! :)

**Comment by dkayiwa** [comment]
@manuela-grindei my advice is that you do not miss out on the chance to learn this. As long as you plan to have a software development career, there are things you will have no option but patiently lean, regardless of the project you will chose to work on! :)

**Pull request closed without merging**